File: scripts/docker-api.ts
/**
 * Docker Management API for SentinelKarma
 * Control and monitor Docker services programmatically
 */

import { exec, execSync, spawn } from 'child_process';
import { promisify } from 'util';
import * as fs from 'fs';
import * as path from 'path';

const execAsync = promisify(exec);

// Types
interface CommandResult {
  success: boolean;
  stdout?: string;
  stderr?: string;
  returncode?: number;
}

interface ServiceStatus {
  [key: string]: 'running' | 'stopped' | 'unknown';
}

interface HealthCheck {
  [service: string]: {
    status: string;
    api?: string;
    mqtt?: string;
    details?: any;
  };
}

interface DockerStats {
  docker?: string;
  log_server?: any;
  data?: {
    contract_data: number;
    malicious_logs: number;
    logs_normal: number;
  };
}

interface MonitorOptions {
  full?: boolean;
  verbose?: boolean;
  mute?: boolean;
}

interface OverburstOptions {
  methods?: string[];
  rate?: number;
  errorRate?: number;
  baselineLatency?: number;
  burstLatency?: number;
  seconds?: number;
  parallel?: boolean;
}

/**
 * Docker Manager API for SentinelKarma services
 */
export class DockerManagerAPI {
  private projectDir: string;
  private managerScript: string;
  private services: string[] = [
    'mosquitto',
    'agent',
    'generator',
    'saver',
    'log-server'
  ];

  constructor(projectDir: string = '/home/water/SentinelKarma') {
    this.projectDir = projectDir;
    this.managerScript = path.join(projectDir, 'manager.sh');
  }

  /**
   * Run shell command
   */
  private async runCommand(
    cmd: string[],
    cwd?: string
  ): Promise<CommandResult> {
    try {
      const cmdStr = cmd.join(' ');
      const { stdout, stderr } = await execAsync(cmdStr, {
        cwd: cwd || this.projectDir
      });
      return {
        success: true,
        stdout,
        stderr
      };
    } catch (error: any) {
      return {
        success: false,
        stdout: error.stdout,
        stderr: error.stderr,
        returncode: error.code
      };
    }
  }

  /**
   * Run manager.sh with flags
   */
  private async runManager(flags: string[]): Promise<CommandResult> {
    return this.runCommand(['bash', this.managerScript, ...flags]);
  }

  /**
   * Run docker compose command
   */
  private async runDockerCompose(args: string[]): Promise<CommandResult> {
    return this.runCommand(['docker', 'compose', ...args]);
  }

  /**
   * Check if all dependencies are installed
   */
  async checkDependencies(): Promise<{ allPresent: boolean; output: string }> {
    const result = await this.runManager(['--check']);
    return {
      allPresent: result.success,
      output: result.stdout || ''
    };
  }

  /**
   * Install required dependencies
   */
  async installDependencies(): Promise<boolean> {
    const result = await this.runManager(['--install']);
    return result.success;
  }

  /**
   * Setup Docker and build images
   */
  async setupDocker(): Promise<boolean> {
    const result = await this.runManager(['--docker']);
    return result.success;
  }

  /**
   * Start services
   */
  async startServices(services?: string[]): Promise<boolean> {
    let result: CommandResult;
    if (services) {
      result = await this.runDockerCompose(['up', '-d', ...services]);
    } else {
      result = await this.runDockerCompose(['up', '-d']);
    }
    return result.success;
  }

  /**
   * Stop services
   */
  async stopServices(services?: string[]): Promise<boolean> {
    let result: CommandResult;
    if (services) {
      result = await this.runDockerCompose(['stop', ...services]);
    } else {
      result = await this.runManager(['--stop']);
    }
    return result.success;
  }

  /**
   * Restart a service
   */
  async restartService(service: string): Promise<boolean> {
    const result = await this.runDockerCompose(['restart', service]);
    return result.success;
  }

  /**
   * Get service status
   */
  async getServiceStatus(service?: string): Promise<ServiceStatus> {
    const args = service ? ['ps', service] : ['ps'];
    const result = await this.runDockerCompose(args);
    
    const services: ServiceStatus = {};
    
    if (result.success && result.stdout) {
      const lines = result.stdout.split('\n');
      
      for (const line of lines.slice(1)) {
        if (line.trim()) {
          const parts = line.split(/\s+/);
          if (parts.length >= 3) {
            const name = parts[0];
            const status = line.includes('Up') ? 'running' : 'stopped';
            services[name] = status;
          }
        }
      }
    }
    
    return services;
  }

  /**
   * Get logs from a service
   */
  async getLogs(service: string, lines: number = 100): Promise<string> {
    const result = await this.runDockerCompose([
      'logs',
      '--tail',
      lines.toString(),
      service
    ]);
    return result.stdout || '';
  }

  /**
   * Start monitoring
   */
  async monitorStart(options: MonitorOptions = {}): Promise<CommandResult> {
    const flags = ['--monitor'];
    
    if (options.full) flags.push('--full');
    if (options.verbose) flags.push('--verbose');
    if (options.mute) flags.push('--mute');
    
    return this.runManager(flags);
  }

  /**
   * Start test mode
   */
  async testMode(): Promise<boolean> {
    const result = await this.runManager(['--test']);
    return result.success;
  }

  /**
   * Run overburst test
   */
  async overburst(options: OverburstOptions = {}): Promise<boolean> {
    const env = { ...process.env };
    
    if (options.methods) {
      env.OVER_METHODS = options.methods.join(' ');
    }
    if (options.rate) {
      env.OVER_RATE = options.rate.toString();
    }
    if (options.errorRate) {
      env.OVER_ERR = options.errorRate.toString();
    }
    if (options.baselineLatency) {
      env.OVER_BASE_LAT = options.baselineLatency.toString();
    }
    if (options.burstLatency) {
      env.OVER_BURST_LAT = options.burstLatency.toString();
    }
    if (options.seconds) {
      env.OVER_SECS = options.seconds.toString();
    }
    if (options.parallel !== undefined) {
      env.OVER_PARALLEL = options.parallel ? '1' : '0';
    }
    
    try {
      execSync(`bash ${this.managerScript} --overburst`, {
        cwd: this.projectDir,
        env
      });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Build Docker images
   */
  async buildImages(): Promise<boolean> {
    const result = await this.runDockerCompose(['build']);
    return result.success;
  }

  /**
   * Purge project containers and images
   */
  async purgeProject(): Promise<boolean> {
    const result = await this.runManager(['--docker-purge']);
    return result.success;
  }

  /**
   * Execute command in service container
   */
  async execCommand(service: string, command: string): Promise<CommandResult> {
    return this.runDockerCompose(['exec', '-T', service, 'sh', '-c', command]);
  }

  /**
   * Check health of all services
   */
  async healthCheck(): Promise<HealthCheck> {
    const health: HealthCheck = {};
    
    // Get service status
    const status = await this.getServiceStatus();
    
    for (const service of this.services) {
      health[service] = {
        status: status[`sentinelkarma-${service}-1`] || 'unknown'
      };
    }
    
    // Special health checks
    try {
      // Log server health
      const response = await fetch('http://localhost:9000/health');
      if (response.ok) {
        health['log-server'].api = 'healthy';
        health['log-server'].details = await response.json();
      } else {
        health['log-server'].api = 'unreachable';
      }
    } catch {
      health['log-server'].api = 'unreachable';
    }
    
    // Mosquitto health
    const mqttResult = await this.execCommand(
      'mosquitto',
      "mosquitto_sub -h localhost -t '$SYS/#' -C 1"
    );
    health['mosquitto'].mqtt = mqttResult.success ? 'healthy' : 'unhealthy';
    
    return health;
  }

  /**
   * Get statistics
   */
  async getStats(): Promise<DockerStats> {
    const stats: DockerStats = {};
    
    // Docker stats
    const dockerResult = await this.runDockerCompose(['stats', '--no-stream']);
    if (dockerResult.success) {
      stats.docker = dockerResult.stdout;
    }
    
    // Log server stats
    try {
      const response = await fetch('http://localhost:9000/stats');
      if (response.ok) {
        stats.log_server = await response.json();
      }
    } catch {
      // Ignore
    }
    
    // Data directory stats
    const dataDir = path.join(this.projectDir, 'data');
    if (fs.existsSync(dataDir)) {
      const countFiles = (dir: string, pattern: string): number => {
        const fullPath = path.join(dataDir, dir);
        if (!fs.existsSync(fullPath)) return 0;
        return fs.readdirSync(fullPath)
          .filter(f => f.endsWith(pattern))
          .length;
      };
      
      stats.data = {
        contract_data: countFiles('contract_data', '.log'),
        malicious_logs: countFiles('malicious_logs', '.jsonl'),
        logs_normal: countFiles('logs_normal', '.jsonl')
      };
    }
    
    return stats;
  }

  /**
   * Stream logs in real-time
   */
  streamLogs(service: string, callback: (data: string) => void): void {
    const child = spawn('docker', ['compose', 'logs', '-f', service], {
      cwd: this.projectDir
    });
    
    child.stdout.on('data', (data) => {
      callback(data.toString());
    });
    
    child.stderr.on('data', (data) => {
      callback(`[ERROR] ${data.toString()}`);
    });
  }

  /**
   * Wait for service to be healthy
   */
  async waitForService(
    service: string,
    healthCheckUrl?: string,
    timeout: number = 30
  ): Promise<boolean> {
    const startTime = Date.now();
    
    while (Date.now() - startTime < timeout * 1000) {
      const status = await this.getServiceStatus(service);
      const serviceName = `sentinelkarma-${service}-1`;
      
      if (status[serviceName] === 'running') {
        if (healthCheckUrl) {
          try {
            const response = await fetch(healthCheckUrl);
            if (response.ok) {
              return true;
            }
          } catch {
            // Continue waiting
          }
        } else {
          return true;
        }
      }
      
      // Wait 1 second before next check
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
    
    return false;
  }
}

// Example usage
if (require.main === module) {
  (async () => {
    console.log('Docker Manager API Examples');
    console.log('='.repeat(60));
    
    // Initialize API
    const docker = new DockerManagerAPI();
    
    // Check dependencies
    console.log('\n1. Checking dependencies...');
    const deps = await docker.checkDependencies();
    console.log(`   Dependencies OK: ${deps.allPresent}`);
    
    // Get service status
    console.log('\n2. Service Status:');
    const status = await docker.getServiceStatus();
    for (const [service, state] of Object.entries(status)) {
      console.log(`   ${service}: ${state}`);
    }
    
    // Health check
    console.log('\n3. Health Check:');
    const health = await docker.healthCheck();
    for (const [service, info] of Object.entries(health)) {
      console.log(`   ${service}: ${info.status}`);
    }
    
    // Get stats
    console.log('\n4. Statistics:');
    const stats = await docker.getStats();
    if (stats.data) {
      console.log(`   Contract data files: ${stats.data.contract_data}`);
      console.log(`   Malicious logs: ${stats.data.malicious_logs}`);
      console.log(`   Normal logs: ${stats.data.logs_normal}`);
    }
    
    console.log('\n' + '='.repeat(60));
    console.log('Quick Usage:');
    console.log(`
import { DockerManagerAPI } from './docker-api';

// Initialize
const docker = new DockerManagerAPI();

// Start services
await docker.startServices();

// Start monitoring (full mode)
await docker.monitorStart({ full: true });

// Get logs
const logs = await docker.getLogs('agent', 50);

// Restart a service
await docker.restartService('log-server');

// Check health
const health = await docker.healthCheck();
    `);
  })();
}

File: scripts/docker_api.py
#!/usr/bin/env python3
"""
Docker Management API for SentinelKarma
Control and monitor Docker services programmatically
"""

import subprocess
import json
import time
from typing import List, Dict, Optional, Any
from pathlib import Path


class DockerManagerAPI:
    """Docker Manager API for SentinelKarma services"""
    
    def __init__(self, project_dir: str = "/home/water/SentinelKarma"):
        """
        Initialize Docker Manager API
        
        Args:
            project_dir: Path to SentinelKarma project
        """
        self.project_dir = Path(project_dir)
        self.manager_script = self.project_dir / "manager.sh"
        
        # Service names
        self.services = [
            "mosquitto",
            "agent", 
            "generator",
            "saver",
            "log-server"
        ]
    
    def _run_command(self, cmd: List[str], cwd: str = None) -> Dict:
        """Run shell command and return result"""
        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                cwd=cwd or self.project_dir,
                check=True
            )
            return {
                "success": True,
                "stdout": result.stdout,
                "stderr": result.stderr
            }
        except subprocess.CalledProcessError as e:
            return {
                "success": False,
                "stdout": e.stdout,
                "stderr": e.stderr,
                "returncode": e.returncode
            }
    
    def _run_manager(self, flags: List[str]) -> Dict:
        """Run manager.sh with given flags"""
        cmd = ["bash", str(self.manager_script)] + flags
        return self._run_command(cmd)
    
    def _run_docker_compose(self, args: List[str]) -> Dict:
        """Run docker compose command"""
        cmd = ["docker", "compose"] + args
        return self._run_command(cmd)
    
    def check_dependencies(self) -> Dict:
        """Check if all dependencies are installed"""
        result = self._run_manager(["--check"])
        return {
            "all_present": result["success"],
            "output": result["stdout"]
        }
    
    def install_dependencies(self) -> bool:
        """Install required dependencies"""
        result = self._run_manager(["--install"])
        return result["success"]
    
    def setup_docker(self) -> bool:
        """Setup Docker and build images"""
        result = self._run_manager(["--docker"])
        return result["success"]
    
    def start_services(self, services: List[str] = None) -> bool:
        """
        Start specific services or all
        
        Args:
            services: List of service names (default: all)
        """
        if services is None:
            result = self._run_docker_compose(["up", "-d"])
        else:
            result = self._run_docker_compose(["up", "-d"] + services)
        return result["success"]
    
    def stop_services(self, services: List[str] = None) -> bool:
        """
        Stop specific services or all
        
        Args:
            services: List of service names (default: all)
        """
        if services is None:
            result = self._run_manager(["--stop"])
        else:
            result = self._run_docker_compose(["stop"] + services)
        return result["success"]
    
    def restart_service(self, service: str) -> bool:
        """Restart a specific service"""
        result = self._run_docker_compose(["restart", service])
        return result["success"]
    
    def get_service_status(self, service: str = None) -> Dict:
        """
        Get status of service(s)
        
        Args:
            service: Service name (default: all)
        """
        if service:
            result = self._run_docker_compose(["ps", service])
        else:
            result = self._run_docker_compose(["ps"])
        
        if result["success"]:
            # Parse output
            lines = result["stdout"].strip().split('\n')
            services = {}
            
            for line in lines[1:]:  # Skip header
                if line.strip():
                    parts = line.split()
                    if len(parts) >= 3:
                        name = parts[0]
                        status = "running" if "Up" in line else "stopped"
                        services[name] = status
            
            return services
        return {}
    
    def get_logs(self, service: str, lines: int = 100) -> str:
        """
        Get logs from a service
        
        Args:
            service: Service name
            lines: Number of lines to return
        """
        result = self._run_docker_compose(["logs", "--tail", str(lines), service])
        return result["stdout"] if result["success"] else ""
    
    def monitor_start(self, full: bool = False, verbose: bool = False, mute: bool = False) -> Dict:
        """
        Start monitoring
        
        Args:
            full: Enable full mode (auto-mint)
            verbose: Show all telemetry
            mute: Run in background
        """
        flags = ["--monitor"]
        if full:
            flags.append("--full")
        if verbose:
            flags.append("--verbose")
        if mute:
            flags.append("--mute")
        
        result = self._run_manager(flags)
        return result
    
    def test_mode(self) -> bool:
        """Start test mode with collectors"""
        result = self._run_manager(["--test"])
        return result["success"]
    
    def overburst(self, methods: List[str] = None, rate: int = 2000) -> bool:
        """
        Run overburst test
        
        Args:
            methods: RPC methods to burst
            rate: Request rate
        """
        import os
        env = os.environ.copy()
        
        if methods:
            env["OVER_METHODS"] = " ".join(methods)
        env["OVER_RATE"] = str(rate)
        
        cmd = ["bash", str(self.manager_script), "--overburst"]
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            cwd=self.project_dir,
            env=env
        )
        return result.returncode == 0
    
    def build_images(self) -> bool:
        """Rebuild Docker images"""
        result = self._run_docker_compose(["build"])
        return result["success"]
    
    def purge_project(self) -> bool:
        """Purge all project containers and images"""
        result = self._run_manager(["--docker-purge"])
        return result["success"]
    
    def exec_command(self, service: str, command: str) -> Dict:
        """
        Execute command in a service container
        
        Args:
            service: Service name
            command: Command to execute
        """
        result = self._run_docker_compose(["exec", "-T", service, "sh", "-c", command])
        return result
    
    def health_check(self) -> Dict:
        """Check health of all services"""
        health = {}
        
        # Check each service
        status = self.get_service_status()
        for service in self.services:
            health[service] = {
                "status": status.get(f"sentinelkarma-{service}-1", "unknown")
            }
        
        # Special health checks
        import requests
        
        # Log server health
        try:
            response = requests.get("http://localhost:9000/health", timeout=2)
            if response.status_code == 200:
                health["log-server"]["api"] = "healthy"
                health["log-server"]["details"] = response.json()
        except:
            health["log-server"]["api"] = "unreachable"
        
        # Mosquitto health
        result = self.exec_command("mosquitto", "mosquitto_sub -h localhost -t '$SYS/#' -C 1")
        health["mosquitto"]["mqtt"] = "healthy" if result["success"] else "unhealthy"
        
        return health
    
    def get_stats(self) -> Dict:
        """Get statistics from all services"""
        stats = {}
        
        # Docker stats
        result = self._run_docker_compose(["stats", "--no-stream"])
        if result["success"]:
            stats["docker"] = result["stdout"]
        
        # Log server stats
        try:
            import requests
            response = requests.get("http://localhost:9000/stats", timeout=2)
            if response.status_code == 200:
                stats["log_server"] = response.json()
        except:
            pass
        
        # Data directory stats
        data_dir = self.project_dir / "data"
        if data_dir.exists():
            stats["data"] = {
                "contract_data": len(list((data_dir / "contract_data").glob("*.log"))) if (data_dir / "contract_data").exists() else 0,
                "malicious_logs": len(list((data_dir / "malicious_logs").glob("*.jsonl"))) if (data_dir / "malicious_logs").exists() else 0,
                "logs_normal": len(list((data_dir / "logs_normal").glob("*.jsonl"))) if (data_dir / "logs_normal").exists() else 0,
            }
        
        return stats


# Example usage
if __name__ == "__main__":
    # Initialize API
    docker = DockerManagerAPI()
    
    print("Docker Manager API Examples")
    print("=" * 60)
    
    # Check dependencies
    print("\n1. Checking dependencies...")
    deps = docker.check_dependencies()
    print(f"   Dependencies OK: {deps['all_present']}")
    
    # Get service status
    print("\n2. Service Status:")
    status = docker.get_service_status()
    for service, state in status.items():
        print(f"   {service}: {state}")
    
    # Health check
    print("\n3. Health Check:")
    health = docker.health_check()
    for service, info in health.items():
        print(f"   {service}: {info.get('status', 'unknown')}")
    
    # Get stats
    print("\n4. Statistics:")
    stats = docker.get_stats()
    if "data" in stats:
        print(f"   Contract data files: {stats['data']['contract_data']}")
        print(f"   Malicious logs: {stats['data']['malicious_logs']}")
        print(f"   Normal logs: {stats['data']['logs_normal']}")
    
    print("\n" + "=" * 60)
    print("Quick Usage:")
    print("""
from docker_api import DockerManagerAPI

# Initialize
docker = DockerManagerAPI()

# Start services
docker.start_services()

# Start monitoring (full mode)
docker.monitor_start(full=True)

# Get logs
logs = docker.get_logs("agent", lines=50)

# Restart a service
docker.restart_service("log-server")

# Check health
health = docker.health_check()
    """)